internals/overlord/checkstate/handlers.go

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) 2024 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package checkstate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/overlord/state"
+	"github.com/canonical/pebble/internals/plan"
+)
+
+func (m *CheckManager) doPerformCheck(task *state.Task, tomb *tomb.Tomb) error {
+	var details checkDetails
+	m.state.Lock()
+	changeID := task.Change().ID()
+	err := task.Get("check-details", &details)
+	m.state.Unlock()
+	if err != nil {
+		return fmt.Errorf("cannot get check details for task %q: %v", task.ID(), err)
+	}
+
+	m.planLock.Lock()
+	plan := m.plan
+	m.planLock.Unlock()
+	config, ok := plan.Checks[details.Name]
+	if !ok {
+		logger.Debugf("Check %q no longer exists in plan.", details.Name)
+		return nil
+	}
+
+	m.checksLock.Lock()
+	data := m.checks[config.Name]
+	if data == nil {
+		data = &checkData{}
+		m.checks[config.Name] = data
+	}
+	data.config = config
+	data.cancel = func() { tomb.Kill(nil) }
+	data.changeID = changeID
+	m.checksLock.Unlock()
+
+	logger.Debugf("Performing check %q with period %v", details.Name, config.Period.Value)
+	ticker := time.NewTicker(config.Period.Value)
+	defer ticker.Stop()
+
+	chk := newChecker(config, plan)
+	for {
+		select {
+		case <-ticker.C:
+			err := runCheck(tomb.Context(nil), config, chk, config.Timeout.Value)
+			if err != nil {
+				// Check failed, switch to recovering a failing check.
+				failures, actionRan := m.handleCheckFailure(config, err)
+
+				// Hold checksLock while updating state so that if someone
+				// calls Checks() they always see the latest state. We must
+				// always acquire checksLock and then state lock in that order.
+
+				m.checksLock.Lock()
+
+				m.state.Lock()
+				newChangeID := m.recoverCheckChange(config)
+				m.state.EnsureBefore(0)
+				m.state.Unlock()
+
+				data.failures = failures
+				data.actionRan = actionRan
+				data.changeID = newChangeID
+				m.checksLock.Unlock()
+
+				// Returning the error means perform-check goes to Error status
+				// and logs the error to the task log.
+				return err
+			}
+
+		case <-tomb.Dying():
+			logger.Debugf("Check %q stopped during perform-check: %v", config.Name, tomb.Err())
```

Comment:
I think that if you ```cancel``` this due to a plan change, this error will be ```nil```, so the message will always end with ```<nil>``` in those cases ?

Reply:
Yes, that's right. I figured `<nil>` was probably okay for a debug message, but I've improved this now.